addons/crm/tests/test_crm_statusbar_time.py

```diff
@@ -0,0 +1,212 @@
+            _update_statustime_tracking(track_json_statustime3, [('-', self.stage_2)])
+            self.assertEqual(json.loads(opp_time1.json_statustime), track_json_statustime1)
+            self.assertEqual(json.loads(opp_time2.json_statustime), track_json_statustime2)
+            self.assertEqual(json.loads(opp_time3.json_statustime), track_json_statustime3)
+
+            now.return_value += timedelta(minutes=10)
+            batch.stage_id = self.stage_3
+            _update_statustime_tracking(track_json_statustime1, [('+', self.stage_1)])
+            _update_statustime_tracking(track_json_statustime2, [('+', self.stage_1)])
+            _update_statustime_tracking(track_json_statustime3, [('+', self.stage_1)])
+            self.assertEqual(json.loads(opp_time1.json_statustime), track_json_statustime1)
+            self.assertEqual(json.loads(opp_time2.json_statustime), track_json_statustime2)
+            self.assertEqual(json.loads(opp_time3.json_statustime), track_json_statustime3)
+
+            now.return_value += timedelta(minutes=10)
+            batch.stage_id = self.stage_2
+            self.assertEqual(json.loads(opp_time1.json_statustime), track_json_statustime1)
+            self.assertEqual(json.loads(opp_time2.json_statustime), track_json_statustime2)
+            self.assertEqual(json.loads(opp_time3.json_statustime), track_json_statustime3)
+
+            now.return_value += timedelta(minutes=10)
+            batch.stage_id = self.stage_4
+            _update_statustime_tracking(track_json_statustime1, [('+', self.stage_2)])
+            _update_statustime_tracking(track_json_statustime2, [('+', self.stage_2)])
+            _update_statustime_tracking(track_json_statustime3, [('+', self.stage_2)])
+            self.assertEqual(json.loads(opp_time1.json_statustime), track_json_statustime1)
+            self.assertEqual(json.loads(opp_time2.json_statustime), track_json_statustime2)
+            self.assertEqual(json.loads(opp_time3.json_statustime), track_json_statustime3)
+
+            now.return_value += timedelta(minutes=10)
+            batch.stage_id = self.stage_2
+            _update_statustime_tracking(track_json_statustime1, [('-', self.stage_2)])
+            _update_statustime_tracking(track_json_statustime2, [('-', self.stage_2)])
+            _update_statustime_tracking(track_json_statustime3, [('-', self.stage_2)])
+            self.assertEqual(json.loads(opp_time1.json_statustime), track_json_statustime1)
+            self.assertEqual(json.loads(opp_time2.json_statustime), track_json_statustime2)
+            self.assertEqual(json.loads(opp_time3.json_statustime), track_json_statustime3)
+
+    def test_queries_batch_json_statustime(self):
+        """When change the stage in a batch, the mixin should only query in batch to reduce queries
+           Formula for calculating number of queries is 11 + 7 * (number of leads)"""
```

Comment:
How come it is so many when updating one record?
Don't we expected to fetch all tracking information in one query,  and setting the new json in another?
Imagine if we reorganize and drop a column with 100 leads inside?

Reply:
@flch-odoo 
In master, changing the stage of one lead is done in 17 queries and changing the stage of two leads is done in 21 queries ; so this works as intended and only adds the one query with the sorted browse for a batch
The queries are for example for creating a `mail.message` records, a `mail.tracking.value` record, ...